libp2p/peerstore.nim

```diff
@@ -31,35 +30,55 @@ type
   # Each book contains a book (map) and event handler(s)
   AddressBook* = object
     book*: Table[PeerID, HashSet[MultiAddress]]
-    addrChange: AddrChangeHandler
+    changeHandlers: seq[AddrChangeHandler]
+  
+  ProtoBook* = object
+    book*: Table[PeerID, HashSet[string]]
+    changeHandlers: seq[ProtoChangeHandler]
+  
+  KeyBook* = object
+    book*: Table[PeerID, PublicKey]
+    changeHandlers: seq[KeyChangeHandler]
+  
+  MetadataBook* = object
+    book*: Table[PeerID, Table[string, seq[byte]]]
```

Comment:
Is this completely arbitrary? What do we imagine storing here? (Generally seems useful, mostly curious)

Reply:
Yes, it's supposed to be a general purpose key-value metadata store per peer. In my mind we can use it to keep performance related metrics, such as peer age, throughput, bandwidth, etc. In the go- and js implementations it's also used to store peer location, libp2p agent version and other peer-related metadata determined during the `identify` procedure (see e.g. https://github.com/libp2p/js-libp2p/pull/800).

Reply:
I'm not 100% sure about `seq[byte]` as it can just hold pure data, no references etc.. but for now it's just fine, it's something that can be fixed when we figure out better.

Reply:
> I noticed no raises which is good and so we can probably add on top {.push raises: [Defect].}

Ah, great catch, lets get that added right at the top of the file @jm-clius!

Reply:
In fact, to clarify - there rule of thumb is that we use this style everywhere possible, the only place/reason it can't be used is in the presence of async, non-async apis should default to this style.

Reply:
(off topic, obviously this was meant to go under https://github.com/status-im/nim-libp2p/pull/505#issuecomment-766785882, but when I do "Quote reply" it puts it under the wrong thread 😨. I've noticed github doing this twice already - user error or bug?)

Reply:
> Is this completely arbitrary? What do we imagine storing here? (Generally seems useful, mostly curious)

Good point. I think there will be many metadata types and it's probably better to have a parameterizable type here like:

```nim
  MetadataBook*[T] = object
    book*: Table[PeerID, Table[string, T]]
```

Another approach is to have a base `PeerBookEntry*[T]` type that everything inherits from and then simply define specialized `PeerBookEntry*[T]` per book. So for example, 

```nim
type 
  PeerBookEntry*[T] = object of RootObj
    entry*: T

  AddressEntry* = object of PeerBookEntry[MultiAddress]

  AddressBook* = object
    book*: Table[PeerID, HashSet[AddressEntry]]
    addrChange: AddrChangeHandler
    changeHandlers: seq[AddrChangeHandler]
```

Then use that to build arbitrary metadata entries...

Reply:
Ok, great suggestions and I especially like the consistency of the second suggestion. To be clear, do you then suggest that specialized entry objects be defined for `AddressEntry`, `ProtoEntry` and `KeyEntry`, but that the `MetadataBook` maintains the generic `PeerBookEntry[T]` so peer book clients can define their own entry types?

Reply:
I like the generic entry yes!, maybe codegen will be bloaty but that's not our fault (nim things)

Reply:
> Ok, great suggestions and I especially like the consistency of the second suggestion. To be clear, do you then suggest that specialized entry objects be defined for AddressEntry, ProtoEntry and KeyEntry, but that the MetadataBook maintains the generic PeerBookEntry[T] so peer book clients can define their own entry types?

Yep, precisely. In fact, we can make `MetadataBook` be the base for the more specialised book types as well. The only thing I'm not content with is the implicit `entry*: T` in `PeerBookEntry`, I wish Nim had a more concise syntax for that sort of thing.

Reply:
Yup, this is a very reasonable suggestion, since addresses, keys and protos can be seen as specialised types of metadata in any case.

My only concern would be that the API should still make explicit the subtle differences between the books. For example, an AddressBook maps a `peerId` to a _collection_ of `addrs` (and therefore should allow clients to `add` individual `addrs` to each collection), whereas a KeyBook only allows a single `PublicKey` entry per `peerId` (and therefore only allows `set`ting a singular entry).

I've attempted to do the above in 12e0716. This significantly changes (and simplifies) the `PeerStore` by making extensive use of generics. Each `PeerStore` still consists of a specialised `addressBook`, `keyBook` and `protoBook`, but now also contains one "any-purpose" `metadataBook` of generic type `PeerBook[T]` (this can easily be changed to a `seq` of such any-purpose books if there is a need for extensibility). Of course, this does not really allow for run-time object variance in the `metadataBook` (which I don't think generics are suitable for, in any case?), but does allow clients to specify their own metadata `type` when instantiating a new `PeerStore`. I'm not exactly happy with the `PeerStore` itself being genericised, but since it requires a concrete type when constructing the `metadataBook` I don't really see a way around this. I'll be glad for any suggestions to make this even more flexible. cc @sinkingsugar @dryajov

Reply:
Honestly, the concept of a free form metadata store doesn't make much sense in this context and I would just leave it out, allowing anyone who needs additional information to just extend the `PeerBook` and `PeerStores` accordingly.